### PR TITLE
fix: injection detection bug

### PR DESCRIPTION
**Pull request by cartcrom**
## Description

Mini portfolio PR caused a regression in how we detect injected wallet that do not inject right away. This PR moves `Connection` fields that have dynamic value based on injection state to be getter functions rather than static vars.

This also adds more tests for edge cases for combinations of different browser environments / injections

#### Manual

_[Steps of how you are testing the change and ensuring no regression.]_

#### Automated
- [x] Unit test
- [ ] Integration/E2E test

### CODE REVIEW

File: src/components/WalletModal/index.tsx [hunk 1]
```diff
@@ -108,39 +107,33 @@ export default function WalletModal({ openSettings }: { openSettings: () => void
...
-  ])
+  }, [connectedWallets, addWalletToConnectedWallets, lastActiveWalletAddress, account, connector, chainId, provider])
 
   // Used to track the state of the drawer in async function
   const drawerOpenRef = useRef(drawerOpen)
   drawerOpenRef.current = drawerOpen
 
   const tryActivation = useCallback(
     async (connection: Connection) => {
+      // Skips wallet connection if the connection should override the default behavior, i.e. install metamask or launch coinbase app
```
Comment (zzmp) [inline src/components/WalletModal/index.tsx:134]: This is new behavior. Why was this added? Is it related to the bug?

File: src/connection/utils.ts [hunk 2]
```diff
@@ -1,15 +1,21 @@
-export const isInjected = Boolean(window.ethereum)
+import { isMobile } from 'utils/userAgent'
```
Comment (zzmp) [inline src/connection/utils.ts:1]: Can you minimize risk by mapping the logic 1:1, instead of changing the abstraction (ie introducing `isMobile` in this file, and changing the mobile logic in `src/connection/index.ts`)

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/connection/index.ts [hunk 3]
```diff
@@ -50,73 +49,71 @@ const [web3Network, web3NetworkHooks] = initializeConnector<Network>(
...
-export const darkInjectedConnection: Connection = {
-  ...baseInjectedConnection,
-  icon: isGenericInjector ? INJECTED_DARK_ICON_URL : METAMASK_ICON_URL,
-}
-
-export const lightInjectedConnection: Connection = {
-  ...baseInjectedConnection,
-  icon: isGenericInjector ? INJECTED_LIGHT_ICON_URL : METAMASK_ICON_URL,
+  overrideActivate: () => {
+    if (!getIsMetaMaskWallet() && !getIsGenericInjector()) {
```
Comment (zzmp) [inline src/connection/index.ts:77]:
```suggestion
    if (getShouldAdvertiseMetaMask()) {
```

File: src/connection/index.ts [hunk 4]
```diff
@@ -32,13 +30,14 @@ export enum ConnectionType {
...
   connector: Connector
   hooks: Web3ReactHooks
   type: ConnectionType
-  icon?: string
-  shouldDisplay?: boolean
-  overrideActivate?: () => void
+  // TODO(WEB-3130): add darkmode check for icons
+  getIcon?(): string
+  shouldDisplay(): boolean
+  overrideActivate?: () => boolean
```
Comment (zzmp) [inline src/connection/index.ts:40]: The callsite of `overrideActivate` must also be updated.

**Review by zzmp — CHANGES_REQUESTED** [review]

**Review by zzmp — APPROVED** [review]

**Review by cmcewen — APPROVED** [review]
